provider/query_parameter_value.go

```diff
@@ -0,0 +1,52 @@
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Query parameter holds normalized parameter data ready to be inserted in the
+//  final query
+type QueryParameterValue struct {
+	// Token to replace e.g., !TOKEN!
+	Token string
+	// SQL expression to be inserted. Contains "?" that will be replaced with an
+	//  ordinal argument e.g., "$1"
+	SQL string
+	// Value that will be passed to the final query in arguments list
+	Value interface{}
+	// Raw parameter values for debugging and monitoring
+	RawValues map[string]string
+}
+
+// ReplaceParams substitutes configured query parameter tokens for their values
+//  within the provided SQL string
+func ReplaceParams(params map[string]QueryParameterValue, sql string, args *[]interface{}) string {
+	if params == nil {
+		return sql
+	}
+
+	for _, token := range ParameterTokenRegexp.FindAllString(sql, -1) {
+		param := params[token]
+
+		// Replace every ? in the param's SQL with a positional argument
+		paramSQL := ""
```

Comment:
How about using a `strings.Builder` since we build this string char by char?

Better yet, couldn't this just be done with a regular `strings.ReplaceAll`? I assume that `param.SQL` contains at least one `?`. If it doesn't, that should be a configuration error (we should validate that `sql` and `default_sql` contain at least one `?`). And then all you really need is:
```go
*args = append(*args, param.Value)
paramSQL := strings.ReplaceAll(paramSQL, "?", fmt.Sprintf("$%d", len(*args)))
```

Reply:
A `default_sql` value is not allowed to have `?`. I wouldn't require it for `sql` either to support use-cases when a part of a query may or may not be added based on a flag presence.

So we don't know if we should append to `args` before the replacement has happened. Unfortunately, `strings.ReplaceAll` doesn't tell us if it actually did something or not, so in this case we'll have to compare strings before and after the replacement, or check if `?` occurs before that.

This will require additional iteration, so I'd prefer to avoid it.

Reply:
Hm... I didn't think about using params as flags 🤔. I'm not sure if that's a very common use case. And even then, you should be able to do it even if you require `sql` to have `?` (`default_sql` should actually be required to NOT have `?`, my bad). You could have e.g. `sql: "...whatever... AND ? = ?"`. I'd make `?` mandatory in `sql` and keep it simple.

Reply:
Anyway, the point is that in this method we don't know if `param.SQL` was an `sql` or `default_(value|sql)`, so we can't rely on `?` presence (since `defefault_sql` doesn't have it). It means iterating twice through the string(s). I'd prefer to just do replacement "by hand", not a big deal.